packages/contracts-bedrock/contracts/universal/StandardBridge.sol

```diff
@@ -344,18 +362,34 @@ abstract contract StandardBridge {
         bytes calldata _extraData
     ) public onlyOtherBridge returns (bool) {
         try this.completeOutboundTransfer(_localToken, _remoteToken, _to, _amount) {
-            emit ERC20BridgeFinalized(_localToken, _remoteToken, _from, _to, _amount, _extraData);
-            return true;
+            if (_from == address(otherBridge)) {
+                // The _from address is the address of the remote bridge if a transfer fails to be
+                // finalized on the remote chain.
+                // slither-disable-next-line reentrancy-events
+                emit ERC20BridgeRefunded(_localToken, _remoteToken, _to, _amount, _extraData);
+                return true;
+            } else {
+                emit ERC20BridgeFinalized(
+                    _localToken,
+                    _remoteToken,
+                    _from,
+                    _to,
+                    _amount,
+                    _extraData
+                );
+                return true;
+            }
         } catch {
             // Something went wrong during the bridging process, return to sender.
             // Can happen if a bridge UI specifies the wrong L2 token.
-            // We reverse the to and from addresses to make sure the tokens are returned to the
-            // sender on the other chain and preserve the accuracy of accounting based on emitted
+            // Since the _to address never received the L2 token, we set this bridge contract
+            // as the sender address for the refund below, and set the recipient as the _from
+            // address. This preserves the accuracy of accounting based on emitted
             // events.
             _initiateBridgeERC20Unchecked(
```

Comment:
What happens if the return-to-sender fails for some reason? Now both `_from` and `_to` are set to `address(this)` and I guess the funds just get stuck forever?

Reply:
At least with the original design you get a loop that can be resolved by fixing one side or the other

Reply:
We could change the interface of `finalizeBridgeERC20` to include a flag `isRefund` instead of doing this

Reply:
And instead of adding a new event we could just emit the flag as part of the `ERC20BridgeFinalized` and `ERC20BridgeFailed` events, which are both new events and don't need to be backwards-compatible